django/utils/functional.py

```diff
@@ -223,12 +224,40 @@ def keep_lazy(*resultclasses):
     def decorator(func):
         lazy_func = lazy(func, *resultclasses)
 
+        func_params = tuple(inspect.signature(func).parameters.items())
+        # Skip parameters which are *args, or **kwargs, which are intrinsically
+        # multiple arguments and must go through the slow path.
+        contains_var_parameters = bool({
+            k for k, v in func_params
+            if v.kind in (inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD)
+        })
+        non_var_parameters = [
+            k for k, v in func_params
+            if v.kind not in (inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD)
+        ]
+        func_first_param_name = None
+        if non_var_parameters:
+            func_first_param_name = non_var_parameters[0]
```

Comment:
> This section doesn't seem so efficient as we're iterating the parameters twice. It's also a bit difficult to understand.

Yeah, it's a bit of a result of my finding a couple of issues with assumptions baked in the initial version, and unsticking myself from the quagmire of that. Certainly it needs refinement, and at first blush I probably like the variant you've proposed (minor niggle being that I don't like all the functions being called `wrapper`, precisely because it makes reading the cProfile/yappi output more difficult)

I will say I don't care _particularly_ about the efficiency of _generating_ the implementations, because they're only built once per decorated function per process -- I might assume that yours is faster in addition to being neater, but I'm not likely to benchmark either. The performance characteristic is secondary to the readability of the implementation unless it becomes pathologically/noticeably problematic, IMO.